Sources/SourceKitLSP/Rename.swift

```diff
@@ -268,33 +254,146 @@ fileprivate struct SyntacticRenameName {
   }
 }
 
+private extension LineTable {
+  subscript(range: Range<Position>) -> Substring? {
+    guard let start = self.stringIndexOf(line: range.lowerBound.line, utf16Column: range.lowerBound.utf16index),
+      let end = self.stringIndexOf(line: range.upperBound.line, utf16Column: range.upperBound.utf16index)
+    else {
+      return nil
+    }
+    return self.content[start..<end]
+  }
+}
+
+private extension DocumentSnapshot {
+  init(_ url: URL, language: Language) throws {
+    let contents = try String(contentsOf: url)
+    self.init(uri: DocumentURI(url), language: language, version: 0, lineTable: LineTable(contents))
+  }
+}
+
+private extension RenameLocation.Usage {
+  init(roles: SymbolRole) {
+    if roles.contains(.definition) || roles.contains(.declaration) {
+      self = .definition
+    } else if roles.contains(.call) {
+      self = .call
+    } else {
+      self = .reference
+    }
+  }
+}
+
+// MARK: - SourceKitServer
+
+extension SourceKitServer {
+  func rename(_ request: RenameRequest) async throws -> WorkspaceEdit? {
+    let uri = request.textDocument.uri
+    guard let workspace = await workspaceForDocument(uri: uri) else {
+      throw ResponseError.workspaceNotOpen(uri)
+    }
+    guard let languageService = workspace.documentService[uri] else {
+      return nil
+    }
+
+    // Determine the local edits and the USR to rename
+    let renameResult = try await languageService.rename(request)
+    var edits = renameResult.edits
```

Comment:
It’s not quite so simple because `edits` is a `WorkspaceEdit`, which also contains `documentChanges` that can eg. rename a file and I don’t want to drop these from the language service’s `rename` request, eg. if `clangd` returned a workspace edit. But refactored it slightly to remove the force-unwrap.